### DISCUSSION TITLE
How to extract all columns from a query string

### DISCUSSION THREAD

**Discussion by antonylebechec** [duckdb/duckdb#15491] [Ideas]
Is there a way to extract all "needed" columns from a query string?

Let's play with an example:
```
SELECT "#CHROM", "POS", REF, ALT, CAST("FILTER" AS STRING), unnest("hgvs") AS NOMEN
FROM variants
WHERE CAST(ID AS STRING) IS NOT NULL
GROUP BY QUAL
```
I would like this list:
variants."#CHROM"
variants."POS"
variants."REF"
variants."ALT"
variants."FILTER"
variants."hgvs"
variants."ID"
variants."QUAL"

In this example, only 1 table, but possibly multiple tables.
The best way would be to explain/analyse without executing the query (just parsing), and without error when a column or table does not exists (why not a info that identify existing/not existing table/columns).

Do you think it is possible?

**Comment by soerenwolfers** [comment]
Try [`json_serialize_sql` ](https://duckdb.org/docs/data/json/sql_to_and_from_json.html)

For example, using some Python just because it's less messy:

```python
from duckdb import query
import json

q = json.loads(query("""
SELECT json_serialize_sql('SELECT "#CHROM", "POS", REF, ALT, CAST("FILTER" AS STRING), unnest("hgvs") AS NOMEN
FROM variants
WHERE CAST(ID AS STRING) IS NOT NULL
GROUP BY QUAL')
""").df().iloc[0, 0])

def item_generator(json_input):
    if isinstance(json_input, dict):
        if json_input.get('class', None) == 'COLUMN_REF':
            yield from json_input.get('column_names', [])
        for k, v in json_input.items():
            yield from item_generator(v)
    elif isinstance(json_input, list):
        for item in json_input:
            yield from item_generator(item)

list(item_generator(q))
```
```
['#CHROM', 'POS', 'REF', 'ALT', 'FILTER', 'hgvs', 'ID', 'QUAL']
```
